src/librustc_codegen_ssa/mir/block.rs

```diff
@@ -232,12 +232,42 @@ impl<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
             }
 
             mir::TerminatorKind::Return => {
+                if self.fn_ty.variadic {
+                    // FIXME(dlrobertson): We need to find the index of the
+                    // spoofed VaList in the argument list and call va_end
+                    // on the corresponding local ref.
+                    let va_list_idx = self.fn_ty.args.iter().enumerate().find(
+                        |(_, arg)| {
+                            PassMode::Ignore(IgnoreMode::CVarArgs) == arg.mode
+                        }
+                    ).unwrap_or_else(|| {
+                        bug!("C variadic function must have a VaList arg")
+                    });
+
+                    // Given the index of the spoofed VaList in the
+                    // functions args, get the index of the arg
+                    // local reference
+                    #[allow(unused)]
+                    let local_ref_idx = self.locals.indices().nth(
+                        va_list_idx.0
+                    ).unwrap_or_else(|| {
+                        bug!("C variadic fns must have 1 or more args")
+                    });
+
+                    // Get the arg local reference and call va_end
```

Comment:
So you haven't actually touched the MIR part of the compiler all, and this is rather backend-agnostic codegen, which I'm quite unfamiliar with to be honest. I see your problem though, I don't think it's solved by modifying `librustc_mir` at all (though I could be wrong). Have you tried using `codegen_place` (on ? It probably won't work, but worth a shot. Otherwise, I think the solution is to add a field like `valist_slot: Option<PlaceRef<'tcx, Bx::Value,>>,` to the `FunctionCx` struct (and use `Option::replace`/`Option::take` on it) to store `tmp` from line 590 (where you generate the slot for the `VaList`) and then you can get it back here.

Reply:
>  I think the solution is to add a field like valist_slot: Option<PlaceRef<'tcx, Bx::Value,>>, to the FunctionCx struct

After thinking about it more, I think this is the way to go. I'll think about it more, but I think I'll implement this tomorrow

Reply:
FWIW I got this working. still got some cleanup to do, but we're now injecting `va_end` correctly before `ret`! Now on to figuring out `call`...